// @arkecosystem/crypto-identities/dist/keys.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Keys = void 0;
const bcrypto_1 = require("bcrypto");
// @ts-ignore
const wif_1 = __importDefault(require("wif"));
const errors_1 = require("./errors");
const hash_algorithms_1 = require("./hash-algorithms");
class Keys {
    static fromPassphrase(passphrase, compressed = true) {
        return Keys.fromPrivateKey(hash_algorithms_1.HashAlgorithms.sha256(Buffer.from(passphrase, "utf8")), compressed);
    }
    static fromPrivateKey(privateKey, compressed = true) {
        privateKey = privateKey instanceof Buffer ? privateKey : Buffer.from(privateKey, "hex");
        return {
            publicKey: bcrypto_1.secp256k1.publicKeyCreate(privateKey, compressed).toString("hex"),
            privateKey: privateKey.toString("hex"),
            compressed,
        };
    }
    static fromWIF(wif, options) {
        const { version, compressed, privateKey } = wif_1.default.decode(wif, options.wif);
        if (version !== options.wif) {
            throw new errors_1.NetworkVersionError(options.wif, version);
        }
        return {
            publicKey: bcrypto_1.secp256k1.publicKeyCreate(privateKey, compressed).toString("hex"),
            privateKey: privateKey.toString("hex"),
            compressed,
        };
    }
}
exports.Keys = Keys;
